apps/src/applab/RecordListener.js

```diff
@@ -0,0 +1,204 @@
+  /**
+   * @type {{string}}
+   * @private
+   */
+  this.tableName_ = tableName;
+
+  /**
+   *  @type {function}
+   *  @private
+   */
+  this.callback_ = callback;
+
+  /**
+   * @typedef {Object.<number, string>} IdToJsonMap Map from record id
+   * to a JSON-stringified copy of the corresponding record.
+   */
+
+  /**
+   * IdToJsonMap representing the contents of the table.
+   * We use this to efficiently test whether a record's contents has changed.
+   * @type {IdToJsonMap}
+   * @private
+   */
+  this.idToJsonMap_ = {};
+
+  /**
+   * Timeout id of the pending call to window.setTimeout.
+   * @type {number}
+   * @private
+   */
+  this.timeoutId_ = null;
+
+  this.scheduleNextFetch_();
+};
+
+/**
+ * @private
+ */
+TableHandler.prototype.scheduleNextFetch_ = function () {
+  this.timeoutId_ = setTimeout(function recordListener() {
+    this.fetchNewRecords_();
+  }.bind(this), RECORD_INTERVAL);
+};
+
+/**
+ * @private
+ */
+TableHandler.prototype.fetchNewRecords_ = function () {
+  var req = new XMLHttpRequest();
+  req.onreadystatechange = this.reportRecords_.bind(this, req);
+  var url = '/v3/shared-tables/' + Applab.channelId + '/' + this.tableName_;
+  req.open('GET', url, true);
+  req.send();
+};
+
+/**
+ * @param {XMLHttpRequest} req
+ * @private
+ */
+TableHandler.prototype.reportRecords_ = function (req) {
+  var done = XMLHttpRequest.DONE || 4;
+  if (req.readyState !== done) {
+    return;
+  }
+
+  if (req.status < 200 || req.status >= 300) {
+    // ignore errors
+    return;
+  }
+
+  if (!this.idToJsonMap_) {
+    // The maps may have been reset while we were fetching the data.
+    return;
+  }
+
+  // Performance notes: on a table with 2000 records on localhost it takes ~1.5 seconds
+  // to fetch the data from dynamodb and 30ms to process it here. Processing time may be
+  // longer on slower machines. One quick and dirty performance improvement is to cache
+  // the responseText here and return if it has not changed (only 1ms). However, a better
+  // performance improvement is to only poll when Pusher tells us to, which would render
+  // the responseText check unnecessary.
+
+  var callback = this.callback_;
+  var records = JSON.parse(req.responseText);
+  var oldIdToJsonMap = this.idToJsonMap_;
+
+  // Look for 'create' and 'update' events, and build the new map of this table.
+  var newIdToJsonMap = {};
+  for (var i = 0; i < records.length; i++) {
+    var record = records[i];
+    var oldJson = oldIdToJsonMap[record.id];
+    var newJson = JSON.stringify(record);
+    if (!oldJson) {
+      callback(record, EventType.CREATE);
+    } else if (oldJson !== newJson) {
+      callback(record, EventType.UPDATE);
+    }
+    newIdToJsonMap[record.id] = newJson;
+  }
+
+  // Look for 'delete' events.
+  for (var oldId in oldIdToJsonMap) {
+    if (!newIdToJsonMap[oldId]) {
+      var deletedRecord = {id: parseInt(oldId, 10)};
+      callback(deletedRecord, EventType.DELETE);
+    }
+  }
```

Comment:
This whole chunk of logic for detecting created/updated/deleted rows is a great candidate for unit tests.  What if you extracted this logic into a static method so you could write simple, stateless tests around it?